lib/sensors/IMU_MadgwickFilter.h

```diff
@@ -0,0 +1,75 @@
+#ifndef _IMU_MADGWICK_FILTER_H_
+#define _IMU_MADGWICK_FILTER_H_
+
+/*
+ * An implementation of the Madgwick AHRS filter.
+ * Fuses IMU measurements to produce orientation quaternions.
+ * Source: https://x-io.co.uk/open-source-imu-and-ahrs-algorithms/
+ */
+
+#include <math.h>
+
+#include <stdexcept>
+
+#define DEFAULT_UPDATE_FREQ 512.0f
+#define DEFAULT_BETA        0.1f
+#define DEFAULT_ZETA        0.0f
+
+#ifndef GYRO_BIAS_DRIFT_COMPENSATION
+#define GYRO_BIAS_DRIFT_COMPENSATION  // define to apply gyro bias drift compensation in filter
+#endif
+
+float invSqrt(float x);
+
+class vec4f {
```

Comment:
if you think vec4f and quaternions may be useful outside of the Madgwick filter it may be good to separate

Reply:
Would vec4f and quaternions get used outside of the Madgwick filter? I sort of intended them just to be used in the filter. I also realized that vec4f and quaternions have pretty much the same implementation, but I wanted to separate them so the filter algorithm was easier to read / cross-reference with the math.

Reply:
👍 Its fine to keep them in there then